### PR TITLE
Rework ABSORBS/ABSORBS_SPLITS behavior

### PR DESCRIPTION
**Pull request by robob27**
#### Summary
Balance "Rework ABSORBS/ABSORBS_SPLITS flags as special attacks with move cost"

#### Purpose of change
Shoggoths feel a bit overtuned as they are currently implemented. 50 hp regen per turn, 400 health and unlimited splits with no cooldown can allow a single shoggoth to make an area just completely off limits. The usual behavior I've seen reported is to just freeze them out of the reality bubble and forget the area completely - making no attempt to fight at all because of the inevitability of being overwhelmed.

#### Describe the solution
- Remove unused "monster_goals" behavior
- Rename "absorb_items" behavior to "ABSORB_ITEMS" to be in line with goal name, matching EAT_CROP definition
- Add "SPLIT" behavior
- The ABSORBS_SPLITS and ABSORBS flag have been removed in favor of 2 separate special attacks that can each have their own cooldowns and many options specified in monster json
- ABSORB_ITEMS and SPLIT can now have a movement cost. The ABSORB_ITEMS movement cost is based on volume and can have minimums and maximums specified.
- The amount of HP gained from absorbing items can now be configured through JSON
- The shoggoth has been modified to have a movement cost for both ABSORB_ITEMS and SPLIT as well as a 5 second cooldown on splitting. The Mythos mod shoggoth was updated so that it should function exactly the same as before.
- It is now possible to add `"invert_result": true | false` to predicates. This allows using the same predicate for both the negative and positive result to drive behavior. For example, the predicate that ensures a monster is not a hallucination could be inverted to ensure that a monster _is_ a hallucination before executing certain behavior. This is used to prioritize splitting over absorbing for monsters with both ABSORB_ITEMS and SPLIT with the "monster_split_possible" predicate.
- Monsters with ABSORB can now have a material array specified to limit the types of materials they are able to absorb. This condition is checked in "monster_items_available" if the material array is specified. A method to return the material_id vector of absorb materials has been added to monster.
- Added a method to check whether a monster has a certain special without regard for if it is on cooldown.
- Added test for new ABSORB_ITEMS/SPLIT special attacks with a shoggoth
- DRY up predicate processing by copy/pasting a for loop that was repeated into a new private method `node_t::process_predicates` that is now called in both places instead.
- Check for a preceding "!" in the "monster_special_available" predicate that allows for checking that the special is available _only_ if the monster actually has that special. This is used to control the behavior for ABSORB_ITEMS such that a monster that also has SPLIT will only try to absorb items when SPLIT is not on cooldown (monsters that have ABSORB_ITEMS but not SPLIT would be unaffected). This prevents monsters with SPLIT/ABSORB_ITEMS from continuously absorbing items if SPLIT is on cooldown, leading to multiple splits in the future.
- Moved special attack documentation to newly created `SPECIAL_ATTACKS.md`. As part of this I manually audited each attack that was documented there. I removed about 3 unused flags and copied everything else that has been implemented as an actual special attack over. This started when I was removing ABSORBS_SPLITS and ABSORBS from there and trying to find where to add documentation for the new special attacks and realized that these seemed to be the last 2 special attack type flags to be migrated.

#### Describe alternatives you've considered
- The PR has evolved a bunch over time based on feedback. I've considered and implemented many alternatives, happy to hear more if they're out there.

#### Testing

Automated testing should cover making sure most of this actually works but here is how to see it in action:
- Spawn a shoggoth and some items
- Kite the shoggoth over the items and ensure the it absorbs them as expected, splitting if fed enough items (6-7 steel frames will do it)
- Try adding some different `absorb_material` options to the shoggoth to test that only those materials are absorbed
- Try configuring different cooldowns for absorbing and splitting. Observe that the shoggoth will only absorb items when it is possible for it to split
- Try spawning a corpse and a flesh eating shoggoth and observe that the corpse contents are left behind
#### Additional context
Initially I set out to add a 2 minute cooldown to the ABSORBS_SPLITS behavior. Kevin pointed out that the behavior didn't have a move cost, Venera3 suggested that a material array would be a good idea, and while working through that I saw an opportunity to migrate what looks to be the last 2 special attack flags to actual special attacks. This was a fun rabbit hole!

### CODE REVIEW

**Comment by Venera3** [comment]
You could comfortably raise the ml-to-hp ratio to 1000 and the cooldown effect's duration to 15 mins. The usual problem spawns are those when a shoggoth noms a microlab's welcoming committee, with each zed corpse triggering a split.

**Comment by robob27** [comment]
@Venera3 Yeah I'd be happy to switch it up to that if that's cool with folks. Truthfully I did want to nerf it a bit more than I did but also didn't want to completely kneecap them... at least all by myself. I feel slightly biased against them 😛.

**Comment by robob27** [comment]
Would it make sense to JSONize the regeneration rate so that it can be customized for future monsters instead of being hardcoded here? This would also allow us to modify the base game shoggoth without force modifying the mod version that appears to exist.

Along the same lines, should a flag exist to ignore this cooldown, which could also be added to the mod version of shoggoth so that it remains unchanged in that way as well, and allows for other potential future monsters that should split more than once per 15 minutes if it makes sense for their balance?

**Comment by Venera3** [comment]
More JSON is always nicer. Both the volume/hp ratio and the hp to split could work very well as JSON fields, and somebody down the line will surely be thankful for a disabling flag.

If you're feeling ambitious, a material array defining what kinds of items they can absorb would be amazing for e.g. corpse-eating critters and similar, and could route all EAT_CROP/EAT_FOOD attacks through ABSORBS_SPLITS (in those cases without splitting, of course). That has the advantage that ABSORBS_SPLITS is handled by monster behaviors, so they actually seek out things to nom.

**Comment by robob27** [comment]
Cool yeah that sounds like fun. On it!

**Comment by kevingranade** [comment]
I think a much more fundamental issue is that the eating and splitting are instantaneous. Even a token move cost is going to nerf the shoggoth as much as or more than the cooldown.

**Comment by robob27** [comment]
@kevingranade Do you think that this would still be worthwhile to merge with the implementation of JSONized cooldown, ml per hp rate etc? Or should this approach be abandoned in favor of adding a move cost to the action? Not sure what all I can accomplish there but happy to try. Also happy to create a separate issue for tackling that later if that works.

**Comment by Venera3** [comment]
While adding a move cost makes sense, it only makes a difference in active combat - most of the problem isn't them splitting in combat, but a random spawn in the reality bubble going ham on items unopposed and filling a whole room/subway tunnel with wall-to-wall shoggies. That would be nicely limited by the higher volume/hp and hp/split ratios, and the cooldown+move cost certainly helps when you have to engage them.

**Comment by robob27** [comment]
I think by adding the movecost we might be able to reduce the cooldown. I'm still testing a bit but I'm working on adding a different move cost for absorbing and splitting. The result should be that it takes a while for a single shoggoth to spawn a bunch of splits from a big pile of materials.  But yeah I think adding both things (in JSON of course) allows a lot of room for more tight balance on this sort of stuff per monster and some interesting design space. I'll see what I can cook up 😄

**Comment by robob27** [comment]
To address the specific problem of shoggoths getting out of control when they are in the reality bubble but not being actively engaged, maybe in another PR it would be possible to make the behavior only occur when hostile? I assume (maybe wrongly) that this monster has the goal consume_items even if not actively engaged in combat. Maybe it could be limited to something it does when it feels like it needs to deal with a threat. This could still occur without the player potentially but maybe would help limit it from getting insane without the player?

**Comment by robob27** [comment]
Converting to draft for now as I still need to work on the material array. I think I'm getting there on the movement cost stuff though. Tried to record a video but it was too big (I'll work on that) but can kinda see from the log:

image

It takes some time for the shoggoth to absorb all of the material and make all the splits.

**Comment by robob27** [comment]
I think maybe the absorb move cost should be based on the volume of the consumed object rather than just a static cost? A shoggoth should probably be able to swallow up a pen in very little time but take longer for a steel frame.

**Comment by robob27** [comment]
Getting very close now. We have:

- consume_items action migrated to special attack in line with eat_crops allowing simplification of monmove logic (just calls special attacks based on action)
- move cost for absorb and splitting based on volume consumed with the ability to specify min and max values
- JSONized pretty much all variables the absorb/split behavior
- ability to specify a material array for what materials a monster can absorb (even leaves behind corpse loot when consuming flesh)

One known issue right now is that specifying any sort of cooldown on absorb/split via the special attack system causes the monster to stop absorbing items after the first item is absorbed as opposed to when the split occurs (as I intend to make it). I think it might make sense to separate ABSORBS from the SPLITS behavior fully - having a special attack that deals with absorbing, and another that deals with splitting. Ideally I could make it so that they could both have independent cooldowns if desired (of course with possibility of no cooldown on both), with the "split" goal taking higher priority over the "absorb" goal (so ideally it will attempt to split as soon as it has the right amount of HP instead of continuing to absorb). Will continue on this tomorrow but any feedback on everything so far is super welcome.

**Comment by robob27** [comment]
I think this is ready for review. The PR description has been updated with a new summary of all of the changes. Tests ran successfully locally 🎉.

**Comment by robob27** [comment]
It occurred to me that I should write a test for monsters with absorb only and I found an issue with my modifications to the "monster_special_available" predicate through the test (yay tests). Working on this now. Everything else is still good for feedback though, just particularly the absorb only monster behavior is flawed until my next commit.

**Comment by robob27** [comment]
Build error seems like a flakey test, I don't think it's related to these changes.

File: src/monattack.cpp [hunk 1]
```diff
@@ -339,6 +339,99 @@ bool mattack::eat_crop( monster *z )
     return true;
 }
 
+bool mattack::split( monster *z )
+{
+    if( !( z->get_hp() / 2 > z->type->hp ) ) {
+        return false;
+    }
+
```
Comment (kevingranade) [inline src/monattack.cpp:347]: This check is redundant.
```suggestion
```

File: src/monattack.cpp [hunk 2]
```diff
@@ -339,6 +339,99 @@ bool mattack::eat_crop( monster *z )
...
+}
+
+bool mattack::absorb_items( monster *z )
+{
+    map &here = get_map();
+
+    if( !here.has_items( z->pos() ) ) {
+        return false;
+    }
+
```
Comment (kevingranade) [inline src/monattack.cpp:377]: Also redundant, loops are just skipped if the target square is empty.
```suggestion
```

**Pull request merged**